crypto/provider_core.c

```diff
@@ -824,18 +795,46 @@ int ossl_provider_forall_loaded(OSSL_LIB_CTX *ctx,
     OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CONFIG, NULL);
 #endif
 
-    if (store != NULL) {
-        provider_activate_fallbacks(store);
-
-        CRYPTO_THREAD_read_lock(store->lock);
-        /*
-         * Now, we sweep through all providers
-         */
-        ret = provider_forall_loaded(store, NULL, cb, cbdata);
+    if (store == NULL)
+        return 1;
+    provider_activate_fallbacks(store);
 
+    /*
+     * Under lock, grab a copy of the provider list and up_ref each
+     * provider so that they don't disappear underneath us.
+     */
+    CRYPTO_THREAD_read_lock(store->lock);
+    provs = sk_OSSL_PROVIDER_dup(store->providers);
+    if (provs == NULL) {
         CRYPTO_THREAD_unlock(store->lock);
+        return 0;
     }
+    j = sk_OSSL_PROVIDER_num(provs);
+    for (i = 0; i < j; i++)
+        if (!ossl_provider_up_ref(sk_OSSL_PROVIDER_value(provs, i)))
+            goto err_unlock;
+    CRYPTO_THREAD_unlock(store->lock);
 
+    /*
+     * Now, we sweep through all providers not under lock
+     */
+    for (j = 0; j < i; j++) {
+        OSSL_PROVIDER *prov = sk_OSSL_PROVIDER_value(provs, j);
+
+        if (prov->flag_activated && !cb(prov, cbdata))
```

Comment:
Is the cb ever likely to do ossl_provider_free()? What happens in that scenario?

Reply:
If the callback _owns_ a reference to the provider, it is entitled to free it.  It shouldn't free the reference that this function has --> it doesn't own it.  If the callback frees the provider (and no other references exist), the provider will persist until this function frees it later.

This isn't a case I'm too concerned about, the ref counting is more to avoid threading issues.

It would be possible to more eagerly free the local references to providers, but it doesn't seem all that useful.

Reply:
I was thinking of the case where the for_all is used to perform a free of its objects..

Reply:
That works.  They hang around until this function frees them after the callbacks are all done.